NNAPI: Support const values in binary ops

axitkhurana opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #59434

Summary:

NNAPI converter failed with 1 const value and one tensor earlier
Code suggestions from @dreiss

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary_const

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

Commit:
NNAPI: Support const values in binary ops

Summary:

NNAPI converter failed with 1 const value and one tensor earlier

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

Commit:
Update on "NNAPI: Support const values in binary ops"


Summary:

NNAPI converter failed with 1 const value and one tensor earlier

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

[ghstack-poisoned]

Commit:
Update on "NNAPI: Support const values in binary ops"


Summary:

NNAPI converter failed with 1 const value and one tensor earlier

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

Commit:
Update on "NNAPI: Support const values in binary ops"


Summary:

NNAPI converter failed with 1 const value and one tensor earlier

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

[ghstack-poisoned]

Commit:
Update on "NNAPI: Support const values in binary ops"


Summary:

NNAPI converter failed with 1 const value and one tensor earlier

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

Commit:
Update on "NNAPI: Support const values in binary ops"


Summary:

NNAPI converter failed with 1 const value and one tensor earlier

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

Commit:
Update on "NNAPI: Support const values in binary ops"


Summary:

NNAPI converter failed with 1 const value and one tensor earlier

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

Commit:
Update on "NNAPI: Support const values in binary ops"


Summary:

NNAPI converter failed with 1 const value and one tensor earlier

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

Commit:
Update on "NNAPI: Support const values in binary ops"


Summary:

NNAPI converter failed with 1 const value and one tensor earlier

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

[ghstack-poisoned]

Commit:
Update on "NNAPI: Support const values in binary ops"


Summary:

NNAPI converter failed with 1 const value and one tensor earlier

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

[ghstack-poisoned]

Commit:
Update on "NNAPI: Support const values in binary ops"


Summary:

NNAPI converter failed with 1 const value and one tensor earlier

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

Commit:
Update on "NNAPI: Support const values in binary ops"


Summary:

NNAPI converter failed with 1 const value and one tensor earlier

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

Commit:
Update on "NNAPI: Support const values in binary ops"


Summary:

NNAPI converter failed with 1 const value and one tensor earlier

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

Commit:
Update on "NNAPI: Support const values in binary ops"


Summary:

NNAPI converter failed with 1 const value and one tensor earlier
Code from @dreiss

https://gist.github.com/dreiss/c5a0a6c093c3b798e4c1a4bc3c79b3ce

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary_const

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

[ghstack-poisoned]

Commit:
Update on "NNAPI: Support const values in binary ops"


Summary:

NNAPI converter failed with 1 const value and one tensor earlier
Code suggestions from @dreiss

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary_const

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

[ghstack-poisoned]

Commit:
Update on "NNAPI: Support const values in binary ops"


Summary:

NNAPI converter failed with 1 const value and one tensor earlier
Code suggestions from @dreiss

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary_const

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

[ghstack-poisoned]

Commit:
Update on "NNAPI: Support const values in binary ops"


Summary:

NNAPI converter failed with 1 const value and one tensor earlier
Code suggestions from @dreiss

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary_const

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

dreiss commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -469,11 +472,14 @@ def get_tensor_operand_by_jitval_fixed_size(self, jitval):
-            operand_id = self.add_tensor_operand_for_weight(value)
+            # assume constants are contiguous
+            if dim_value == DimOrder.CHANNELS_LAST:
+                value.permute(0, 2, 3, 1)
```
This is not an in-place op.  I think you need to do `value = value.permute(0, 2, 3, 1)`.

Should probably first update the test so it fails due to this.

dreiss commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -469,11 +472,14 @@ def get_tensor_operand_by_jitval_fixed_size(self, jitval):
         return op_id, oper
 
-    def get_tensor_operand_or_constant(self, jitval):
+    def get_tensor_operand_or_constant(self, jitval, dim_value=DimOrder.PRESUMED_CONTIGUOUS):
```
Why call this `dim_value` instead of `dim_order?

dreiss commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -1230,9 +1236,14 @@ def _do_add_binary(self, node, opcode, fuse_code, *, qparams=None):
-        in0_id, in0_oper = self.get_tensor_operand_by_jitval(node.inputsAt(0))
-        in1_id, in1_oper = self.get_tensor_operand_by_jitval(node.inputsAt(1))
+        if self.has_operand_for_jitval(node.inputsAt(0)):
+            in0_id, in0_oper = self.get_tensor_operand_by_jitval_fixed_size(node.inputsAt(0))
```
Does this need to be fixed size?  I think I had this in my version because I'm based on a revision where we're forcing fixed size here.

dreiss commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -534,8 +540,8 @@ def compute_operand_shape(self, op_id, dim, expr):
     def transpose_to_nhwc(self, in_id, oper):
-        if oper.shape[2:] != (1, 1):
-            raise Exception("Automatic transpose only supported for H,W == 1,1")
+        if oper.dim_order != DimOrder.UNKNOWN_CONSTANT and oper.shape[2:] != (1, 1):
```
Why was this change needed?

Commit:
Update on "NNAPI: Support const values in binary ops"


Summary:

NNAPI converter failed with 1 const value and one tensor earlier
Code suggestions from @dreiss

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary_const

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

[ghstack-poisoned]

axitkhurana commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -469,11 +472,14 @@ def get_tensor_operand_by_jitval_fixed_size(self, jitval):
-            operand_id = self.add_tensor_operand_for_weight(value)
+            # assume constants are contiguous
+            if dim_order == DimOrder.CHANNELS_LAST:
+                value = value.permute(0, 2, 3, 1).flatten().reshape(value.shape)
```
Is there a better way to do this? If I don't flatten and reshape, broadcasting is a problem with tensor shapes mismatched

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

Commit:
Update on "NNAPI: Support const values in binary ops"


Summary:

NNAPI converter failed with 1 const value and one tensor earlier
Code suggestions from @dreiss

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary_const

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

dreiss commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -418,6 +418,8 @@ def add_tensor_operand_for_weight(self, tensor):
             buf_num,
             offset,
             tsize))
+        if dim_order == DimOrder.CHANNELS_LAST:
```
Can you just leave a comment here?  Something like

Because the NNAPI operand will have shape NHWC, we need to permute the torch tensor to the same dimension order to ensure that the data will be properly laid out after we call contiguous on it.

Commit:
Update on "NNAPI: Support const values in binary ops"


Summary:

NNAPI converter failed with 1 const value and one tensor earlier
Code suggestions from @dreiss

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary_const

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).

Commit:
Update on "NNAPI: Support const values in binary ops"


Summary:

NNAPI converter failed with 1 const value and one tensor earlier
Code suggestions from @dreiss

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_pointwise_binary_const

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28893881](https://our.internmc.facebook.com/intern/diff/D28893881)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28893881).
